### PR TITLE
Keras support - add cosine proximity criterion

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

add cosine proximity criterion

## How was this patch tested?

Local test, merged test with zhichao's unit test utility

### CODE REVIEW

**Comment by wzhongyuan** [comment]
@PR passed after refinement based on 1.2
http://172.168.2.101:8080/job/BigDL-PR-V/177/

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializerSpec.scala [hunk 1]
```diff
@@ -2058,6 +2058,20 @@ class ModuleSerializerSpec extends FlatSpec with Matchers {
     loadedModel.isTraining() should be (false)
   }
 
+  "HardSigmoid serialization" should "work properly" in {
+    val hardSigmoid = HardSigmoid()
```
Comment (zhichao-li) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializerSpec.scala]: Wondering should we expose this as a generic test and put it into `beforeEach` for every `Layer` test? `BigDLSpecHelper` or `BigDLLayerSpec` would be a good place for this.

Comment (zhichao-li) [reply spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializerSpec.scala]: @yiheng any suggestion? I think torchspec would like this as well.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CosineProximityCriterion.scala [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      override def apply(v1: Array[T], v2: Int): Unit = {
+        loss = ev.plus(ev.negative(v1(v2)), loss)
+      }
+    }
+    DenseTensorApply.apply1(mul, func)
+    loss = ev.divide(loss, ev.fromType(mul.nElement()))
+    loss
+  }
+
+  private def l2Norm(input : Tensor[T]) : Tensor[T] = {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CosineProximityCriterion.scala]: Looks like a good utility method. not sure if we can move it to Tensor class. or maybe we can put it into a following PR.

**Comment by zhichao-li** [comment]
Should we add `(val sizeAverage: Boolean = true)` as well? as all of the existing Criterion do support this option. Another thing is Keras1.2.2 only support `sizeAverage=true` and Keras2 support `sizeAverage=false`.  Maybe we can just stick with Keras1.2.2 in this PR.

**Comment by zhichao-li** [comment]
just wondering how do we differentiate batch input and unbatch input? do we need to add some doc saying we support batch input only?

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializerSpec.scala [hunk 3]
```diff
@@ -2058,6 +2058,20 @@ class ModuleSerializerSpec extends FlatSpec with Matchers {
...
+    ModulePersister.saveToFile("/tmp/hardSigmoid.bigdl", hardSigmoid, true)
+    val loadedModel = ModuleLoader.loadFromFile("/tmp/hardSigmoid.bigdl")
+
+    val input = Tensor(2, 2).rand()
+
+    val res1 = hardSigmoid.forward(input)
+
+    val res2 = hardSigmoid.forward(input)
```
Comment (zhichao-li) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializerSpec.scala]: should be `loadedModule` here?

**Comment by wzhongyuan** [comment]
@zhichao-li   For serialization unit test, I do have some thoughts to refine it so that we can easily check if all layers have been covered,

**Comment by wzhongyuan** [comment]
@zhichao-li   both batch and un-batched input are supported, Already tested it.

**Comment by wzhongyuan** [comment]
@zhichao-li  Any other concerns ?

**Comment by zhichao-li** [comment]
add some docs/comments for: "both batch and un-batched input are supported"?
LGTM except for this

**Comment by wzhongyuan** [comment]
@zhichao-li  added comment and also run another PR successfully 
http://172.168.2.101:8080/job/BigDL-PR-V/238/

**Pull request merged**